core/shared/src/main/scala/fs2/async/immutable/Signal.scala

```diff
@@ -36,23 +37,91 @@ abstract class Signal[F[_], A] { self =>
   def get: F[A]
 }
 
-object Signal {
+object Signal extends SignalLowPriorityImplicits {
+
+  implicit def applicativeInstance[F[_]](
+      implicit effectEv: Effect[F],
+      ec: ExecutionContext
+  ): Applicative[Signal[F, ?]] =
+    new Applicative[Signal[F, ?]] {
+      override def map[A, B](fa: Signal[F, A])(f: A => B): Signal[F, B] = Signal.map(fa)(f)
+
+      override def pure[A](x: A): Signal[F, A] = fs2.async.mutable.Signal.constant(x)
+
+      override def ap[A, B](ff: Signal[F, A => B])(fa: Signal[F, A]): Signal[F, B] =
+        new Signal[F, B] {
+
+          override def discrete: Stream[F, B] =
+            nondeterministicZip(ff.discrete, fa.discrete).map { case (f, a) => f(a) }
+
+          override def continuous: Stream[F, B] = Stream.repeatEval(get)
+
+          override def get: F[B] = ff.get.ap(fa.get)
+        }
+    }
+
+  private def nondeterministicZip[F[_], A0, A1](xs: Stream[F, A0], ys: Stream[F, A1])(
```

Comment:
Just remark I would add this in separate PR + tests for it

Reply:
As a side note to `yip` derived from  this implementation. This does not handle any scopes, errors, recoveries etc. (I hope with signal it is safe to assume they are not issue).

 In Stream version however, this will have these to be addressed, and likely we will need to use merge-like pattern, unless new `Fold` will allow us to do something better.

Reply:
I don't really understand how scopes work (and I don't know in particular what's missing here scope-wise) so I'll need help there if I were to implement it :(, but I can try reading through the code some more. I'm not sure what you mean by errors and recovery though. As far I can tell this stream will fail if any of its constituent streams fail, and recovery never comes for free anyways; the standard `bracket` and restart pattern should work here as well.

Reply:
@changlinli no worries, lets focus on that in separate PRs. Thanks for this work.